### ISSUE TITLE
Why can not convert ssd_mobilnetv3 model file tflite format?

### ISSUE THREAD

**Issue by HKVINS89757** [tensorflow/models#7873]
I have trained the mobilnetv3 -ssd model 

The test program runs well on the computer

But it throw errors when I convert it to 'tflite' format!!!

I obtained tflite_graph.pb with 'export_tflite_ssd_graph.py'

Then I try to use 'toco tool' to convert it to tflite file, but it throw errors. like follows picture.

![QQ图片20191128155208](https://user-images.githubusercontent.com/42164696/69787287-2430e580-11f7-11ea-94ae-97061053ecca.png)

It says "Check failed: model.HasArray(input) || model.optional_arrays.count(input) Input: FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1 missing for op: FeatureExtractor/MobilenetV3/expanded_conv/add "

I try to find node named "FeatureExtractor/MobilenetV3/expanded_conv/add"

The result like follows fig:
![QQ图片20191128155628111111111](https://user-images.githubusercontent.com/42164696/69787506-a8836880-11f7-11ea-9433-01cec1a2f20e.png)

One of it inputs named: "FeatureExtractor/MobilenetV3/expanded_conv/input" 

Then I find its definition as follows :
![QQ图片201911281558592222222222](https://user-images.githubusercontent.com/42164696/69787710-01530100-11f8-11ea-9539-5bb28754e0a3.png)

look at the above ,its input is named  "FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1" !

PS:' Toco ' convert tool giving a perfect result when I dealing with mobilnetv2-ssd model.

**Comment by HKVINS89757** [comment]
中文：

我使用了最新版的tensorlfow-models 训练了mobilnetv3-ssd。

训练的模型在电脑上可以正常使用，

但是，移到嵌入式端的时侯需要tflite文件

首先我先使用 export_tflite_ssd_graph.py 把ckpt文件转换成 tflite_graph.pb

然后再用toco转换成嵌入式端可以使用的tflite文件，但是转换的时侯出现了错误
具体来说就是 如下图红框框的部分
![QQ图片20191128155208](https://user-images.githubusercontent.com/42164696/69788789-2fd1db80-11fa-11ea-8814-fc2031c267bf.png)
主要提示就是  FeatureExtractor/MobilenetV3/expanded_conv/add 这个操作不能操作 FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1 这个节点

那么跟进FeatureExtractor/MobilenetV3/expanded_conv/add  这个操作看下，如下图：
![QQ图片20191128155628111111111](https://user-images.githubusercontent.com/42164696/69788888-5c85f300-11fa-11ea-9c4b-27f6145c4846.png)
可以看到它可以对 FeatureExtractor/MobilenetV3/expanded_conv/input这个节点操作，再看看 FeatureExtractor/MobilenetV3/expanded_conv/input 这个节点 
![QQ图片201911281558592222222222](https://user-images.githubusercontent.com/42164696/69788987-91924580-11fa-11ea-9d70-0523e6d828f1.png)
可以看到这就是一个Identity操作啊，说明FeatureExtractor/MobilenetV3/expanded_conv/input  跟FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1 就是一回事啊
但是就是toco会不认识这个节点，请问是网络模型的错误还是？？？？？

**Comment by HKVINS89757** [comment]
In addition my training environment：
ubunto 16 04 LTS
tensorflow 1。12

But I think my training environment is well since the model can work well on my computer。

**Comment by HKVINS89757** [comment]
and  my config file  for mobilenet v3：
# SSDLite with Mobilenet v3 large feature extractor.
# Trained on COCO14, initialized from scratch.
# 3.22M parameters, 1.02B FLOPs
# TPU-compatible.
# Users should configure the fine_tune_checkpoint field in the train config as
# well as the label_map_path and input_path fields in the train_input_reader and
# eval_input_reader. Search for "PATH_TO_BE_CONFIGURED" to find the fields that
# should be configured.

model {
  ssd {
    inplace_batchnorm_update: true
    freeze_batchnorm: false
    num_classes: 2
    box_coder {
      faster_rcnn_box_coder {
        y_scale: 10.0
        x_scale: 10.0
        height_scale: 5.0
        width_scale: 5.0
      }
    }
    matcher {
      argmax_matcher {
        matched_threshold: 0.5
        unmatched_threshold: 0.5
        ignore_thresholds: false
        negatives_lower_than_unmatched: true
        force_match_for_each_row: true
        use_matmul_gather: true
      }
    }
    similarity_calculator {
      iou_similarity {
      }
    }
    encode_background_as_zeros: true
    anchor_generator {
      ssd_anchor_generator {
        num_layers: 6
        min_scale: 0.2
        max_scale: 0.95
        aspect_ratios: 1.0
        aspect_ratios: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 3.0
        aspect_ratios: 0.3333
      }
    }
    image_resizer {
      fixed_shape_resizer {
        height: 320
        width: 320
      }
    }
    box_predictor {
      convolutional_box_predictor {
        min_depth: 0
        max_depth: 0
        num_layers_before_predictor: 0
        use_dropout: false
        dropout_keep_probability: 0.8
        kernel_size: 3
        use_depthwise: true
        box_code_size: 4
        apply_sigmoid_to_scores: false
        class_prediction_bias_init: -4.6
        conv_hyperparams {
          activation: RELU_6,
          regularizer {
            l2_regularizer {
              weight: 0.00004
            }
          }
          initializer {
            random_normal_initializer {
              stddev: 0.03
              mean: 0.0
            }
          }
          batch_norm {
            train: true,
            scale: true,
            center: true,
            decay: 0.97,
            epsilon: 0.001,
          }
        }
      }
    }
    feature_extractor {
      type: 'ssd_mobilenet_v3_large'
      min_depth: 16
      depth_multiplier: 1.0
      use_depthwise: true
      conv_hyperparams {
        activation: RELU_6,
        regularizer {
          l2_regularizer {
            weight: 0.00004
          }
        }
        initializer {
          truncated_normal_initializer {
            stddev: 0.03
            mean: 0.0
          }
        }
        batch_norm {
          train: true,
          scale: true,
          center: true,
          decay: 0.97,
          epsilon: 0.001,
        }
      }
      override_base_feature_extractor_hyperparams: true
    }
    loss {
      classification_loss {
        weighted_sigmoid_focal {
          alpha: 0.75,
          gamma: 2.0
        }
      }
      localization_loss {
        weighted_smooth_l1 {
          delta: 1.0
        }
      }
      classification_weight: 1.0
      localization_weight: 1.0
    }
    normalize_loss_by_num_matches: true
    normalize_loc_loss_by_codesize: true
    post_processing {
      batch_non_max_suppression {
        score_threshold: 1e-8
        iou_threshold: 0.6
        max_detections_per_class: 100
        max_total_detections: 100
        use_static_shapes: true
      }
      score_converter: SIGMOID
    }
  }
}

train_config: {
  batch_size: 82
  sync_replicas: true
  startup_delay_steps: 0
  replicas_to_aggregate: 32
  fine_tune_checkpoint: "/media/gin/f9a48888-4c41-4b7b-b5e7-7b894f2dea0f/mobilenetv3/1125v3/model.ckpt-26189"
  fine_tune_checkpoint_type:  "detection"
  num_steps: 400000
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  data_augmentation_options {
    ssd_random_crop {
    }
  }
  data_augmentation_options {
    random_vertical_flip {
    }
  }
  data_augmentation_options {
    random_rotation90 {
    }
  }
  optimizer {
    momentum_optimizer: {
      learning_rate: {
        cosine_decay_learning_rate {
          learning_rate_base: 0.4
          total_steps: 400000
          warmup_learning_rate: 0.13333
          warmup_steps: 2000
        }
      }
      momentum_optimizer_value: 0.9
    }
    use_moving_average: false
  }
  max_number_of_boxes: 100
  unpad_groundtruth_tensors: false
}

train_input_reader: {
  tf_record_input_reader {
    input_path: "/media/gin/f9a48888-4c41-4b7b-b5e7-7b894f2dea0f/FrameAbovetheDoorTrain/record1124/train.record"
  }
  label_map_path: "/media/gin/f9a48888-4c41-4b7b-b5e7-7b894f2dea0f/FrameAbovetheDoorTrain/record1124/indoor.pbtxt"
}

eval_config: {
  num_examples: 564
}

eval_input_reader: {
  tf_record_input_reader {
    input_path: "/media/gin/f9a48888-4c41-4b7b-b5e7-7b894f2dea0f/FrameAbovetheDoorTrain/record1124/test.record"
  }
  label_map_path: "/media/gin/f9a48888-4c41-4b7b-b5e7-7b894f2dea0f/FrameAbovetheDoorTrain/record1124/indoor.pbtxt"
  shuffle: false
  num_readers: 1
}

**Comment by HKVINS89757** [comment]
more addition：
ssd -mobilenetv3 model test result on my computer，it is working very well
![index1](https://user-images.githubusercontent.com/42164696/69790520-9278a680-11fd-11ea-8610-144c04c5ce9d.png)

![index3](https://user-images.githubusercontent.com/42164696/69790526-95739700-11fd-11ea-9b3e-02dd4e355d29.png)

**Comment by xjohnxjohn** [comment]
Hi @HKVINS89757 , What the checkpoint you used?

**Comment by HKVINS89757** [comment]
> Hi @HKVINS89757 , What the checkpoint you used?
The name of the model I used for the first training was called ssd_mobilenet_v3_large_coco_2019_0814.
Then，In order to save training time, I used the first trained checkpoint when retraining.

**Comment by HKVINS89757** [comment]
Hi @xjohnxjohn 
,
The name of the model I used for the first training was called ssd_mobilenet_v3_large_coco_2019_0814.
Then，In order to save training time, I used the first trained checkpoint when retraining.

**Comment by xjohnxjohn** [comment]
@HKVINS89757 thank you. I train custom object now, after that convert to tflite and check the result.

**Comment by HKVINS89757** [comment]
@xjohnxjohn Let's solve this problem together! If you have results, please let me know as soon as possible

**Comment by HKVINS89757** [comment]
@xjohnxjohn Have you tested it？

**Comment by xjohnxjohn** [comment]
@HKVINS89757 yes.
Please use bazel to convert with this command:
bazel run -c opt tensorflow/lite/toco:toco -- --input_file=inference_graph/tflite_graph.pb --output_file=inference_graph/detect.tflite --input_shapes=1,320,320,3 --inference_type=FLOAT --mean_values=128 --std_values=128 --change_concat_input_ranges=false --allow_custom_ops --input_arrays='normalized_input_image_tensor' --output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3

**Comment by xjohnxjohn** [comment]
[https://uppic.cc/d/5zqS](url)

**Comment by xjohnxjohn** [comment]
![78598552_3281185655231986_6528398135855677440_o](https://user-images.githubusercontent.com/7420735/70369383-27109200-18eb-11ea-95ec-5a2c3f3b5c4f.jpg)

**Comment by HKVINS89757** [comment]
@xjohnxjohn 
What is your bazel version?

**Comment by xjohnxjohn** [comment]
bazel 1.1.0

**Comment by HKVINS89757** [comment]
@xjohnxjohn   okay   thank you ,   I will  try it again.

**Comment by HKVINS89757** [comment]
@xjohnxjohn
what the paremeter of --allow_custom_ops   ？
NULL？
thank you

**Comment by xjohnxjohn** [comment]
--allow_custom_ops    no parameter

**Comment by mandomwu** [comment]
> 中文：
> 
> 我使用了最新版的tensorlfow-models 训练了mobilnetv3-ssd。
> 
> 训练的模型在电脑上可以正常使用，
> 
> 但是，移到嵌入式端的时侯需要tflite文件
> 
> 首先我先使用 export_tflite_ssd_graph.py 把ckpt文件转换成 tflite_graph.pb
> 
> 然后再用toco转换成嵌入式端可以使用的tflite文件，但是转换的时侯出现了错误
> ps，toco工具可以正常转换mobilnetv2版本的模型。
> 具体来说就是 如下图红框框的部分
> ![QQ图片20191128155208](https://user-images.githubusercontent.com/42164696/69788789-2fd1db80-11fa-11ea-8814-fc2031c267bf.png)
> 主要提示就是 FeatureExtractor/MobilenetV3/expanded_conv/add 这个操作不能操作 FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1 这个节点
> 
> 那么跟进FeatureExtractor/MobilenetV3/expanded_conv/add 这个操作看下，如下图：
> ![QQ图片20191128155628111111111](https://user-images.githubusercontent.com/42164696/69788888-5c85f300-11fa-11ea-9c4b-27f6145c4846.png)
> 可以看到它可以对 FeatureExtractor/MobilenetV3/expanded_conv/input这个节点操作，再看看 FeatureExtractor/MobilenetV3/expanded_conv/input 这个节点
> ![QQ图片201911281558592222222222](https://user-images.githubusercontent.com/42164696/69788987-91924580-11fa-11ea-9d70-0523e6d828f1.png)
> 可以看到这就是一个Identity操作啊，说明FeatureExtractor/MobilenetV3/expanded_conv/input 跟FeatureExtractor/MobilenetV3/Conv/hard_swish/mul_1 就是一回事啊
> 但是就是toco会不认识这个节点，请问是网络模型的错误还是？？？？？
> 
> ps mobilenetv2版本的ssd模型可以正常转成tflite

老哥，关于ssd mobilev3， 我也想做个训练集来训练后转tflite使用；但现在我直接用graph加载下载的pb文件来使用，加载后打印outputs跟你训练后的outputs名字不一样，我这边显示的是：
        boxes = detection_graph.get_tensor_by_name('raw_outputs/box_encodings:0')
        scores = detection_graph.get_tensor_by_name('convert_scores:0')
        classes = detection_graph.get_tensor_by_name('raw_outputs/class_predictions:0')
；是需要什么转换一下么？
另外返回的数据怎么使用啊？怎么把boxes，classes解析出来？能方便指教一下么？ 您微信或qq多少， 不方便留的话加我微信：mandom

**Comment by HKVINS89757** [comment]
@xjohnxjohn
thank you  bro， but I got the seem errors with your command

I guess  it maybe limited bazel version and what‘s your tensorlfow version.

**Comment by HKVINS89757** [comment]
@mandomwu 
已加微信，明天去旅行，周一聊

**Comment by mandomwu** [comment]
thx

**Comment by xjohnxjohn** [comment]
@HKVINS89757 I used OS: ubuntu 16.04LTS, CUDA:9.0, CuDnn:7.6, Tensorflow-gpu==1.12.3.
Can you send your checkpoint to me?

**Comment by HKVINS89757** [comment]
@xjohnxjohn 
Fine,thanks, and what's your email address?

**Comment by xjohnxjohn** [comment]
@HKVINS89757  xjohnmurphy@gmail.com

**Comment by HKVINS89757** [comment]
[v3litemodels1127.zip](https://github.com/tensorflow/models/files/3965873/v3litemodels1127.zip)

Hi,bro

Attached is my model file that needs to be converted

It was converted with  'export_tflite_ssd_graph.py' and the source is a ckpt file

Thank you for your help, waiting for your reply

@xjohnxjohn

**Comment by HKVINS89757** [comment]
@xjohnxjohn 
I sent my checkpoint files to you.
pls Check your email, thanks for you help.
![QQ截图20191217105757](https://user-images.githubusercontent.com/42164696/70961223-1d8de380-20bc-11ea-872e-3fadeff9c0ce.png)

**Comment by HKVINS89757** [comment]
@xjohnxjohn 
Have you tested it?

**Comment by HKVINS89757** [comment]
@xjohnxjohn 
I'm sure that bazel1.1.0 cannot be used to compile tensorflow1.12.3
The bazel1.1.0 can compile tensorflow2.1.0rc2 successfully.

**Comment by HKVINS89757** [comment]
solved。
bazel1.1.0
tensorflow2.1.0rc2

**Comment by ouening** [comment]
> solved。
> bazel1.1.0
> tensorflow2.1.0rc2

请问你是如何使用Tf2.x训练自定义数据集的呢？models库中好些文件都要导入tensorflow.contrib，但是这个在Tf2.x版本中已经移除了的，目前网上的方法使用tensorflow/models训练自定义数据集来做object detection的tf版本好像都是1.x版本的，根本不适应，望赐教！

**Comment by ninetailskim** [comment]
> solved。
> bazel1.1.0
> tensorflow2.1.0rc2

Also have the same problem with @ouening since the contrib has been removed. Could you help me~ Thanks~

**Comment by ravikyram** [comment]
@HKVINS89757 

Please, close this thread if your issue was resolved.Thanks!

**Comment by ravikyram** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
